Assignment3/Q1.py

```diff
@@ -0,0 +1,128 @@
+"""The approach taken in this solution is based on a DFS-like search. The worst case running time
+   is exponential but because of the optimization used with isPrefix(string) method the average
+   running time should be a lot better.
+   The solution given below uses create_corresponding_graph(rows, columns) method to create a graph corresponding
+   to the grid - 2 dimensional array given. It uses a adjacency list representation of a graph
+   (Python dictionary) with an edge between every corresponding two adjacent entries in a given grid.
+   Then using both the given grid and a corresponding graph the DFS-like search is performed on a graph
+   (using words_search_recursive method) but it searches ALL possible paths from each vertex (corresponding
+   to a letter) similar to creating a big search tree with all possible words. Using isPrefix(string) method
+   it 'cuts off' early all the branches of the search tree which is being created that cannot form words."""
+
+from Dictionary import Dictionary
+
+
+def create_corresponding_graph(rows, columns):
```

Comment:
Optimization (if you feel up for it):
Right now you build and store the entire graph, but there might be many cases that are actually never explored. The grid, in a way, is a more compact representation of the same graph. Can we visit the graph without actually needing to store it? Happy to discuss some ideas to point you in the right direction if interested.